### DISCUSSION TITLE
Clarifying API for selection modes for Taps

### DISCUSSION THREAD

**Discussion by handwerkerd** [bokeh/bokeh#13797] [General]
This is based on https://discourse.bokeh.org/t/dynamic-figure-problem-with-bokeh-v-3-4-0-udgrade/11416/6 and it was suggested I post here and ask for feedback from @mattpap.

I contribute to a project that uses `events.Tap` to select elements. When a new element was selected, that selection replaced the previously selected element. With an upgrade from bokeh v.3.3.4 to v3.4.0, every tap results in an additional element being selected and I need to tap each element a second time to unselect. Looking through the v3.4.0 release notes, my best guess is this relates to changing the default tap selection mode from `replace` to `xor` in https://github.com/bokeh/bokeh/pull/13545. That said, that seems to be for `TapTools` and not `events.Tap`. I'm a `bokeh` novice and I'm not completely sure how those interact.

I'm trying to figure out how to set the selection mode for `events.Tap` to `replace` but it's not obvious in the documentation. I'm not sure if this is just something that can be addressed by clarifying documentation or if the mechanisms to change this parameter are not currently accessible through the API. Any help would be appreciated.

If relevant, the bokeh code I'm working with is here: https://github.com/ME-ICA/tedana/blob/f084dc4bf773cf51b52245b5d94ffb5dfebb4339/tedana/reporting/dynamic_figures.py#L411

Thank you

**Comment by bryevdv** [comment]
This is a more specific question that I do think I can answer. The `mode` property is on the `TapTool` itself:

https://docs.bokeh.org/en/latest/docs/reference/models/tools.html#bokeh.models.TapTool.mode

**Reply by handwerkerd** [reply]
Thank you! I was able to work with your answers and I tried to dig a bit more before each new time I posted. I'm far from a bokeh expert and this was a bit tricky to figure out because I ended up needing to change create a `model.TapTool(mode='repeat')` instead of just including the default `tap` in the toolbar.

I suspect https://github.com/bokeh/bokeh/pull/13545 will break functionality in a bunch of other projects and that was far from obvious in the release notes and also a bit tricky for a novice to figure out how to resolve. I had trouble finding any online examples using `model.TapTool` with non-default settings. I eventually did figure this out, but I'll gently prod that this change might benefit from a bit more documentation.

FWIW, the PR that shows how I fixed this for my project is: https://github.com/ME-ICA/tedana/pull/1068
